### PR TITLE
Staged merge of Reed's changes for Publisher Events

### PR DESCRIPTION
**Pull request by fquinner**
https://github.com/OpenMAMA/OpenMAMA/pull/100/files

This is a staging branch which will once again get squash merged to next when it lands.

Something happened to squash the commit history on reed's original branch to include
some of those fixes along with every change which had been submitted on next since the
original fork was taken. In addition, when going through some of these changes, some
changes which weren't on the original pull request have managed to creep in from
somewhere.

Raising this with a mind to create a fresh pull request which will merge safely with
the next branch and with *only the required changes* included.

Signed-off-by: Frank Quinn <frank@openmama.org>

[Review on Reviewable](https://reviewable.io/reviews/openmama/openmama/128)

### CODE REVIEW

File: mama/c_cpp/src/c/listenermsgcallback.c [hunk 1]
```diff
@@ -181,15 +181,7 @@ static void processPointToPointMessage (msgCallback*    callback,
...
-    if (!ctx->mDqContext.mDoNotForward)
-    {
-        mamaSubscription_forwardMsg (self->mSubscription, msg);
-    }
-    else
-    {
-        mama_log (MAMA_LOG_LEVEL_FINER,
-                  "Subscription for %s not forwarded as message seqnum is before seqnum expecting", userSymbol);
-    }
+    mamaSubscription_forwardMsg (self->mSubscription, msg);
```
Comment (fquinner) [inline mama/c_cpp/src/c/listenermsgcallback.c]: Does not look related to change - will take out.

File: mama/c_cpp/src/c/mama/middleware.h [hunk 2]
```diff
@@ -51,7 +51,8 @@ typedef enum mamaMiddleware_
     MAMA_MIDDLEWARE_LBMKOMODO = 15,
     MAMA_MIDDLEWARE_UMDSKOMODO = 16,
     MAMA_MIDDLEWARE_HMS     = 17,
-    MAMA_MIDDLEWARE_MAX     = 18,
+    MAMA_MIDDLEWARE_REFLECT = 18,
+    MAMA_MIDDLEWARE_MAX     = 19,
```
Comment (fquinner) [inline mama/c_cpp/src/c/mama/middleware.h]: Does not look related to change - will take out.

File: mama/c_cpp/src/c/middleware.c [hunk 3]
```diff
@@ -79,10 +79,7 @@ mamaMiddleware_convertFromString (const char*  str)
 
     if (strcasecmp (str, "umdskomodo") == 0)
         return MAMA_MIDDLEWARE_UMDSKOMODO;
-    
-    if (strcasecmp (str, "hms") == 0)
-        return MAMA_MIDDLEWARE_HMS;
-    
+
```
Comment (fquinner) [inline mama/c_cpp/src/c/middleware.c]: Merge seems to have picked this up as an explicit delete - will revert.

**Comment by fquinner** [comment]
Reed, could you review the above intended changes to verify they're OK? As well as that, I also reverted your changes to the following files as they didn't seem to contain any changes relevant to publisher events (or indeed the Windows compatibility changes).

**Comment by fquinner** [comment]
Hit a race condition in MamaPublisherTest.Publish as well, but will not include that as part of this change.

File: mama/c_cpp/src/c/listenermsgcallback.c (hunk 1 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/listenermsgcallback.c]: Agreed.

File: mama/c_cpp/src/c/mama/middleware.h (hunk 2 above)
Comment (reed-jpm-alpert) [inline mama/c_cpp/src/c/mama/middleware.h]: Agreed, that is my test bridge, will submit a pull for that along.

**Comment by reed-jpm-alpert** [comment]
All looks good, thanks.

**Comment by fquinner** [comment]
Now merged into next (d9f874256ceaf7e9cc05d238c68628c36782cf85)
